src/app/clusters/window-covering-server/window-covering-server.cpp

```diff
@@ -0,0 +1,625 @@
+#include <app/util/af-types.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/TypeTraits.h>
+#include <string.h>
+
+#ifdef EMBER_AF_PLUGIN_SCENES
+#include <app/clusters/scenes/scenes.h>
+#endif // EMBER_AF_PLUGIN_SCENES
+
+using namespace chip::app::Clusters;
+
+#define WC_PERCENT100THS_MAX 10000
+
+//------------------------------------------------------------------------------
+// WindowCover
+//------------------------------------------------------------------------------
+
+std::map<chip::EndpointId, WindowCover *> WindowCover::mInstances;
+
+void WindowCover::Init(chip::EndpointId endpoint)
+{
+    // Check if already initialized
+    std::map<chip::EndpointId, WindowCover *>::iterator it = mInstances.find(endpoint);
+    if (it == mInstances.end())
+    {
+        // Initialize endpoint
+        mInstances[endpoint] = new WindowCover(endpoint);
+    }
+}
+
+WindowCover * WindowCover::Get(chip::EndpointId endpoint)
+{
+    std::map<chip::EndpointId, WindowCover *>::iterator it = mInstances.find(endpoint);
+    if (it != mInstances.end())
+    {
+        return it->second;
+    }
+    return nullptr;
+}
+
+void WindowCover::TypeSet(EmberAfWcType type)
+{
+    WindowCovering::Attributes::SetType(mEndpoint, chip::to_underlying(type));
+}
+
+EmberAfWcType WindowCover::TypeGet(void)
+{
+    std::underlying_type<EmberAfWcType>::type value;
+    WindowCovering::Attributes::GetType(mEndpoint, &value);
+    return static_cast<EmberAfWcType>(value);
+}
+
+void WindowCover::ConfigStatusSet(const WindowCover::ConfigStatus & status)
+{
+    uint8_t value = (status.operational ? 0x01 : 0) | (status.online ? 0x02 : 0) | (status.liftIsReversed ? 0x04 : 0) |
+        (status.liftIsPA ? 0x08 : 0) | (status.tiltIsPA ? 0x10 : 0) | (status.liftIsEncoderControlled ? 0x20 : 0) |
+        (status.tiltIsEncoderControlled ? 0x40 : 0);
+    WindowCovering::Attributes::SetConfigStatus(mEndpoint, value);
+}
+
+const WindowCover::ConfigStatus WindowCover::ConfigStatusGet(void)
+{
+    uint8_t value = 0;
+    ConfigStatus status;
+
+    WindowCovering::Attributes::GetConfigStatus(mEndpoint, &value);
+    status.operational             = (value & 0x01) ? 1 : 0;
+    status.online                  = (value & 0x02) ? 1 : 0;
+    status.liftIsReversed          = (value & 0x04) ? 1 : 0;
+    status.liftIsPA                = (value & 0x08) ? 1 : 0;
+    status.tiltIsPA                = (value & 0x10) ? 1 : 0;
+    status.liftIsEncoderControlled = (value & 0x20) ? 1 : 0;
+    status.tiltIsEncoderControlled = (value & 0x40) ? 1 : 0;
+    return status;
+}
+
+void WindowCover::OperationalStatusSet(const WindowCover::OperationalStatus & status)
+{
+    uint8_t global = OperationalStateToValue(status.global);
+    uint8_t lift   = OperationalStateToValue(status.lift);
+    uint8_t tilt   = OperationalStateToValue(status.tilt);
+    uint8_t value  = (global & 0x03) | static_cast<uint8_t>((lift & 0x03) << 2) | static_cast<uint8_t>((tilt & 0x03) << 4);
```

Comment:
We should still add tests for this code, but it's OK to do it in a followup, since nothing actually exercises this codepath with nonzero values for now and figuring out how to do that might take some work.